### PR TITLE
Add support for nodes and edges in geographical space

### PR DESCRIPTION
**Pull request by loreabad6**
See #275 for details. Main issues for now are:

1. the implementation of GeomEdgeSf in geom_edge_sf.R
2. facetting for certain cases
3. handling `sfnetworks` and `sf` in the NAMESPACE

### CODE REVIEW

File: R/tbl_graph.R [hunk 1]
```diff
@@ -25,6 +25,9 @@ create_layout.tbl_graph <- function(graph, layout, circular = FALSE, ...) {
     'layout_ggraph',
     class(layout)
   )
+  if (inherits(graph, 'sfnetwork')) {
+    class(layout) <- append('layout_sfnetwork', class(layout))
+  }
```
Comment (thomasp85) [inline R/tbl_graph.R:30]: Is this classing being used anywhere?

File: R/tbl_graph.R [hunk 2]
```diff
@@ -94,7 +97,8 @@ prepare_graph <- function(graph, layout, direction = 'out', ...) {
     if (!graph_is_treeish) graph <- graph_to_tree(graph, mode = direction)
     graph <- permute(graph, match(seq_len(gorder(graph)), order(node_depth(graph, direction))))
   }
-  as_tbl_graph(graph)
+  if (inherits(graph, "sfnetwork")) graph
+  else as_tbl_graph(graph)
```
Comment (thomasp85) [inline R/tbl_graph.R:101]:
```suggestion
  if (inherits(graph, "sfnetwork")) {
    graph
  } else {
    as_tbl_graph(graph)
  }
```

File: R/layout_sf.R [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#' @importFrom tidygraph as_tbl_graph
+#'
+#' @return A data.frame with the columns `x`, `y`, `circular`.
+#'
+#' @family layout_tbl_graph_*
+layout_tbl_graph_sf <- function(graph, circular = FALSE) {
+  # Check the presence of sf.
+  if (!requireNamespace("sf", quietly = TRUE)) {
+    stop("Package sf required, please install it first.", call. = FALSE)
+  }
```
Comment (thomasp85) [inline R/layout_sf.R:18]:
```suggestion
  check_installed("sf")
```

File: R/layout_sf.R [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  # Check if network is an sfnetwork
+  if (!inherits(graph, "sfnetwork")) {
+    stop("Layout 'sf' needs an 'sfnetwork' graph.", call. = FALSE)
```
Comment (thomasp85) [inline R/layout_sf.R:21]:
```suggestion
    cli::cli_abort("The 'sf' layout needs an {.cls sfnetwork} graph.")
```

File: R/layout_sf.R [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  # Extract X and Y coordinates from the nodes
+  graph <- activate(graph, "nodes")
+  x <- sf::st_coordinates(graph)[,"X"]
+  y <- sf::st_coordinates(graph)[,"Y"]
+  # Create layout data frame
+  nodes <- new_data_frame(list(x = x, y = y))
+  nodes$circular <- FALSE
```
Comment (thomasp85) [inline R/layout_sf.R:29]:
```suggestion
  nodes <- data_frame0(x = x, y = y, circular = FALSE)
```

File: R/layout_sf.R [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  # Convert to tbl_graph to 'unstick' geometry column
+  extra_data <- as_tibble(as_tbl_graph(graph), active = "nodes")
+  warn_dropped_vars(nodes, extra_data)
+  nodes <- cbind(nodes, extra_data[, !names(extra_data) %in% names(nodes), drop = FALSE])
```
Comment (thomasp85) [inline R/layout_sf.R:33]:
```suggestion
  nodes <- combine_layout_nodes(nodes, extra_data)
```

File: R/geom_edge_sf.R [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+#' @format NULL
+#' @usage NULL
+#' @export
+GeomEdgeSf = ggproto("GeomEdgeSf", GeomSf,
+  draw_panel = function(data, panel_params, coords) {
+    names(data) <- sub('edge_', '', names(data))
+    names(data)[names(data) == 'width'] <- 'size'
+    GeomSf$draw_panel(data, panel_params, coords)
+  }
+)
```
Comment (thomasp85) [inline R/geom_edge_sf.R:65]:
```suggestion
GeomEdgeSf = ggproto("GeomEdgeSf", GeomSf,
  draw_panel = function(data, panel_params, coords) {
    names(data) <- sub('edge_', '', names(data))
    names(data)[names(data) == 'width'] <- 'size'
    GeomSf$draw_panel(data, panel_params, coords)
  },
  draw_key = GeomEdgePath$draw_key,
  default_aes = aes(
    edge_colour = 'black', edge_width = 0.5, edge_linetype = 1,
    edge_alpha = NA
  ),
  rename_size = FALSE
)

```

File: R/aaa.R [hunk 8]
```diff
@@ -50,6 +50,23 @@ StatReverse <- ggproto('StatReverse', StatFilter,
...
 
+#' @rdname ggraph-extensions
+#' @format NULL
+#' @usage NULL
+#' @export
+StatFilterSf <- ggproto('StatFilterSf', StatSf,
+  setup_data = function(data, params) {
+    if (any(names(data) == 'filter')) {
+      if (!is.logical(data$filter)) {
+        stop('filter must be logical')
```
Comment (thomasp85) [inline R/aaa.R:61]:
```suggestion
        cli::cli_abort('{.field filter} must be logical')
```

**Review by thomasp85 — CHANGES_REQUESTED** [review]
This is really good. I have made a bunch of suggestions but most are cosmetic

**Comment by thomasp85** [comment]
As for your list of things, I've addressed 1) in my suggestions, 2) is unknown to me - do you have some examples of failures and I'll be happy to look into them, and 3) is also handled in the suggestions I think.

File: R/tbl_graph.R (hunk 1 above)
Comment (loreabad6) [reply R/tbl_graph.R:30]: No, I'll remove it

**Comment by loreabad6** [comment]
Thank you for the suggestions, I have incorporated them. 

Following up on the list:

1. GeomEdgeSf now recognizes aesthetics but does not recognize parameters outside `aes()` unless `edge_*` is added.
```r
ggraph(net, 'sf') +
  geom_node_sf(aes(color = central)) +
  geom_edge_sf(color = 'grey')
#> Warning in layer_sf(geom = GeomEdgeSf, data = data, mapping = mapping, stat = StatFilterSf, :
#> Ignoring unknown parameters: `colour`
```



``` r

### Need to write edge_* to get parameter
ggraph(net, 'sf') +
  geom_node_sf(aes(color = central)) +
  geom_edge_sf(edge_color = 'grey')
```

**Comment by loreabad6** [comment]
And an example of a failed facetting case for point 2:

``` r
library(sfnetworks)
library(tidygraph)
library(ggraph)

net = roxel %>% 
  as_sfnetwork() %>% 
  mutate(centrality = centrality_betweenness()) %>% 
  mutate(central = ifelse(centrality > 1000, T, F)) %>% 
  activate('edges') %>% 
  mutate(azimuth = edge_azimuth(), length = edge_length())

ggraph(net, 'sf') +
  geom_node_sf(color = 'red', size = 0.05) +
  geom_edge_sf(aes(color = type)) +
  facet_graph(type ~ central) 
#> Warning: Unknown or uninitialised column: `.ggraph.index`.
#> Error in `x[i, , drop = drop]`:
#> ! Can't subset rows with `i`.
#> ✖ Logical subscript `i` must be size 1 or 701, not 0.
#> Backtrace:
#>      ▆
#>   1. ├─base::tryCatch(...)
#>   2. │ └─base (local) tryCatchList(expr, classes, parentenv, handlers)
#>   3. │   ├─base (local) tryCatchOne(...)
#>   4. │   │ └─base (local) doTryCatch(return(expr), name, parentenv, handler)
#>   5. │   └─base (local) tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
#>   6. │     └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
#>   7. │       └─base (local) doTryCatch(return(expr), name, parentenv, handler)
#>   8. ├─base::withCallingHandlers(...)
#>   9. ├─base::saveRDS(...)
#>  10. ├─base::do.call(...)
#>  11. ├─base (local) `<fn>`(...)
#>  12. ├─global `<fn>`(input = base::quote("cheap-stoat_reprex.R"))
#>  13. │ └─rmarkdown::render(input, quiet = TRUE, envir = globalenv(), encoding = "UTF-8")
#>  14. │   └─knitr::knit(knit_input, knit_output, envir = envir, quiet = quiet)
#>  15. │     └─knitr:::process_file(text, output)
#>  16. │       ├─knitr:::handle_error(...)
#>  17. │       │ └─base::withCallingHandlers(...)
#>  18. │       ├─base::withCallingHandlers(...)
#>  19. │       ├─knitr:::process_group(group)
#>  20. │       └─knitr:::process_group.block(group)
#>  21. │         └─knitr:::call_block(x)
#>  22. │           └─knitr:::block_exec(params)
#>  23. │             └─knitr:::eng_r(options)
#>  24. │               ├─knitr:::in_input_dir(...)
#>  25. │               │ └─knitr:::in_dir(input_dir(), expr)
#>  26. │               └─knitr (local) evaluate(...)
#>  27. │                 └─evaluate::evaluate(...)
#>  28. │                   └─evaluate:::evaluate_call(...)
#>  29. │                     ├─evaluate (local) handle(...)
#>  30. │                     │ └─base::try(f, silent = TRUE)
#>  31. │                     │   └─base::tryCatch(...)
#>  32. │                     │     └─base (local) tryCatchList(expr, classes, parentenv, handlers)
#>  33. │                     │       └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
#>  34. │                     │         └─base (local) doTryCatch(return(expr), name, parentenv, handler)
#>  35. │                     ├─base::withCallingHandlers(...)
#>  36. │                     ├─base::withVisible(value_fun(ev$value, ev$visible))
#>  37. │                     └─knitr (local) value_fun(ev$value, ev$visible)
#>  38. │                       └─knitr (local) fun(x, options = options)
#>  39. │                         ├─base::withVisible(knit_print(x, ...))
#>  40. │                         ├─knitr::knit_print(x, ...)
#>  41. │                         └─knitr:::knit_print.default(x, ...)
#>  42. │                           └─evaluate (local) normal_print(x)
#>  43. │                             ├─base::print(x)
#>  44. │                             └─ggplot2:::print.ggplot(x)
#>  45. │                               ├─ggplot2::ggplot_build(x)
#>  46. │                               ├─ggraph:::ggplot_build.ggraph(x)
#>  47. │                               ├─base::NextMethod() at ggraph/R/ggraph.R:149:3
#>  48. │                               └─ggplot2:::ggplot_build.ggplot(x)
#>  49. │                                 └─layout$setup(data, plot$data, plot$plot_env)
#>  50. │                                   └─ggplot2 (local) setup(..., self = self)
#>  51. │                                     └─base::lapply(...)
#>  52. │                                       └─ggplot2 (local) FUN(X[[i]], ...)
#>  53. │                                         └─ggraph (local) map_data(...)
#>  54. │                                           └─base::lapply(...) at ggraph/R/facet_graph.R:157:9
#>  55. │                                             └─ggraph (local) FUN(X[[i]], ...)
#>  56. │                                               ├─data[data$.ggraph.index %in% nodes, , drop = FALSE] at ggraph/R/facet_graph.R:158:11
#>  57. │                                               └─sf:::`[.sf`(data, data$.ggraph.index %in% nodes, , drop = FALSE) at ggraph/R/facet_graph.R:158:11
#>  58. │                                                 ├─x[i, , drop = drop]
#>  59. │                                                 └─tibble:::`[.tbl_df`(x, i, , drop = drop)
#>  60. │                                                   └─tibble:::vectbl_as_row_index(i, x, i_arg)
#>  61. │                                                     └─tibble:::vectbl_as_row_location(i, nr, i_arg, assign, call)
#>  62. │                                                       ├─tibble:::subclass_row_index_errors(...)
#>  63. │                                                       │ └─base::withCallingHandlers(...)
#>  64. │                                                       └─vctrs::vec_as_location(...)
#>  65. └─vctrs (local) `<fn>`()
#>  66.   └─vctrs:::stop_indicator_size(...)
#>  67.     └─rlang::cnd_signal(...)
```

File: R/geom_edge_sf.R [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+#'
+geom_edge_sf <- function(mapping = NULL, data = get_sf_edges(),
+                         position = 'identity', show.legend = NA, ...) {
+  mapping <- complete_edge_aes(mapping)
+  mapping <- aes_intersect(mapping, aes(geometry = geometry))
+  c(
+    layer_sf(
+      geom = GeomEdgeSf, data = data, mapping = mapping, stat = StatFilterSf,
+      position = position, show.legend = show.legend, inherit.aes = FALSE,
+      params = list2(na.rm = FALSE, ...)
```
Comment (thomasp85) [inline R/geom_edge_sf.R:41]:
```suggestion
      params = expand_edge_aes(list2(na.rm = FALSE, ...))
```

**Comment by thomasp85** [comment]
The edge param thing was an oversight when I made the last batch of suggestions.

Let me have a look at the facet thing

**Comment by thomasp85** [comment]
So, it seems everything works if `geom_node_sf()` follows the behaviour of the other node geoms and set `data = NULL` rather than `data = get_sf_nodes()`. Would there be any concern with this? It seems all the relevant CRS information is captured in the `geometry` column in the layout so `get_sf_nodes()` is not needed

File: R/geom_node_sf.R [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+#' @examples
+#' if (require("sfnetworks", quietly = TRUE)) {
+#'   gr <- sfnetworks::as_sfnetwork(roxel) %>%
+#'     mutate(centrality = centrality_betweenness())
+#'   ggraph(gr, 'sf') + geom_node_sf(aes(color = centrality))
+#' }
+#'
+#' @export
+#'
+geom_node_sf <- function(mapping = NULL, data = get_sf_nodes(),
```
Comment (thomasp85) [inline R/geom_node_sf.R:35]:
```suggestion
geom_node_sf <- function(mapping = NULL, data = NULL,
```

File: R/geom_node_sf.R [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  )
+}
+
+get_sf_nodes <- function(){
+  function(layout) {
+    nodes <- sf::st_as_sf(attr(layout, "graph"), "nodes")
+    attr(nodes, 'type_ggraph') <- 'node_ggraph'
+    nodes
+  }
+}
```
Comment (thomasp85) [inline R/geom_node_sf.R:54]:
```suggestion

```

**Comment by thomasp85** [comment]
Yeah, I can see the CRS are not picked up with the basic layout as nodes

File: R/geom_node_sf.R [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+      position = position, show.legend = show.legend, inherit.aes = FALSE,
+      params = list2(na.rm = FALSE, ...)
+    ),
+    coord_sf(default = TRUE)
+  )
+}
+
+get_sf_nodes <- function(){
+  function(layout) {
+    nodes <- sf::st_as_sf(attr(layout, "graph"), "nodes")
```
Comment (thomasp85) [inline R/geom_node_sf.R:50]:
```suggestion
    nodes <- sf::st_as_sf(layout)
```

Comment (thomasp85) [reply R/geom_node_sf.R:50]: I believe this should be enough to solve the remaining issues

**Comment by loreabad6** [comment]
Thank you very much for reviewing this @thomasp85, I tested again and seems like all the issues I raised work now.

**Comment by thomasp85** [comment]
Great - can I get you to add a NEWS bullet as well

**Comment by loreabad6** [comment]
> Great - can I get you to add a NEWS bullet as well

Done!

**Comment by thomasp85** [comment]
Thank you!!!

**Pull request merged**